File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  defp numero(0) do
    "No more"
  end

  defp numero(number) do
    rem(number + 100, 100)
  end

  defp bottle_s(1) do
    "bottle"
  end

  defp bottle_s(_) do
    "bottles"
  end

  defp bottle(number) do
    "#{numero(number)} #{bottle_s(number)}"
  end

  defp the_bottle(1) do
    "it"
  end

  defp the_bottle(_) do
    "one"
  end

  defp what_do_you_do(0) do
    "Go to the store and buy some more"
  end

  defp what_do_you_do(number) do
    "Take #{the_bottle(number)} down and pass it around"
  end

  @doc """
  Get a single verse of the beer song
  """

  @spec verse(integer) :: String.t()
  def verse(number) do
    """
    #{bottle(number)} of beer on the wall, #{String.downcase(bottle(number))} of beer.
    #{what_do_you_do(number)}, #{String.downcase(bottle(number - 1))} of beer on the wall.
    """
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range \\ 99..0) do
    range
    |> Enum.map_join("\n",fn x ->
      verse(x)
    end)
  end
end

File: beer-song/src/beer_song.clj
(ns beer-song)
(require '[clojure.core.match :refer [match]])
(require 'clojure.string)

(defn bottle_format
  [n]
  (match n
    0 "no more"
    n n))

(defn plurial
  [n]
  (match n
    1 ""
    _ "s"))

(defn take_bottle
  [n]
  (match n
    1 "it"
    _ "one"))

(defn action
  [n]
  (match n
    0 "Go to the store and buy some more"
    n (str "Take " (take_bottle n) " down and pass it around")))

(defn verse
  "Returns the nth verse of the song."
  [num]
  (let [next_bottle (mod (+ 99 num) 100)]
    (str (clojure.string/capitalize (bottle_format num)) " bottle" (plurial num) " of beer on the wall, " (bottle_format num) " bottle" (plurial num) " of beer.\n" (action num) ", " (bottle_format next_bottle) " bottle" (plurial next_bottle) " of beer on the wall.\n")))

(defn sing
  "Given a start and an optional end, returns all verses in this interval. If
  end is not given, the whole song from start is sung."
  ([start] (sing start 0))
  ([start end] (clojure.string/join "\n" (map #(verse %) (range start (dec end) -1)))))
